tests/stages/test_discovery.py

```diff
@@ -37,7 +37,7 @@ def _caching_for(
     try:
         Output(success=True, message="a test output", hash=hashed_contents).write(
             target_path=Path(path),
-            stage=stage.name,
+            stage=stage_name or "build",
```

Comment:
I know you're just changing logic that already exists, but it's probably a good idea to change this.

This is a dangerous default, as it introduces the possibility of us forgetting to set this value and end up calculating hashes for the build stage when planning the test one.  I'd rather make `stage_name` a mandatory field to prevent bugs.